lib/em-hiredis/new_client.rb

```diff
@@ -0,0 +1,247 @@
+require 'uri'
+
+module EventMachine::Hiredis
+  # Emits the following events
+  #
+  # * :connected - on successful connection or reconnection
+  # * :reconnected - on successful reconnection
+  # * :disconnected - no longer connected, when previously in connected state
+  # * :reconnect_failed(failure_number) - a reconnect attempt failed
+  #     This event is passed number of failures so far (1,2,3...)
+  class BaseClient
+    include EventEmitter
+    include EventMachine::Deferrable
+
+    attr_reader :host, :port, :password, :db
+
+    TRANSITIONS = [
```

Comment:
It also feels messy to have two places where the state machine is defined. Maybe something like:

    state(:connect, [:initial, :connecting], &method(:on_connect))
    ...

Would work better.